### PR TITLE
Add API for vesting and locked balances

### LINKED ISSUE #9846: Add vesting summary

### PR DESCRIPTION
**Pull request by jeremyletang**
close #9846

### CODE REVIEW

File: cmd/data-node/commands/start/sqlsubscribers.go [hunk 1]
```diff
@@ -175,6 +179,7 @@ type SQLSubscribers struct {
 	volumeDiscountStatsSub    *sqlsubscribers.VolumeDiscountStatsUpdated
 	volumeDiscountProgramSub  *sqlsubscribers.VolumeDiscountProgram
 	paidLiquidityFeesStatsSub *sqlsubscribers.PaidLiquidityFeesStats
+	vestinSummarySub          *sqlsubscribers.VestingBalancesSummary
```
Comment (guoguojin) [inline cmd/data-node/commands/start/sqlsubscribers.go:182]:
```suggestion
	vestingSummarySub          *sqlsubscribers.VestingBalancesSummary
```

File: cmd/data-node/commands/start/sqlsubscribers.go [hunk 2]
```diff
@@ -227,6 +232,7 @@ func (s *SQLSubscribers) GetSQLSubscribers() []broker.SQLBrokerSubscriber {
 		s.volumeDiscountStatsSub,
 		s.volumeDiscountProgramSub,
 		s.paidLiquidityFeesStatsSub,
+		// TODO: Add the party locked and vesting balances subscribers here!!
```
Comment (guoguojin) [inline cmd/data-node/commands/start/sqlsubscribers.go:235]:
```suggestion
		s.vestingSummarySub,
```

**Review by guoguojin — CHANGES_REQUESTED** [review]

**Review by guoguojin — DISMISSED** [review]

File: protos/sources/data-node/api/v2/trading_data.proto [hunk 3]
```diff
@@ -849,6 +849,13 @@ service TradingDataService {
     option (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_operation) = {tags: "Volume discount program"};
   }
 
+  // Get vesting balances summary
+  //
+  // Get the information about a party's vesting and locked balances
```
Comment (candida-d) [inline protos/sources/data-node/api/v2/trading_data.proto:854]:
```suggestion
  // Get information about a party's vesting and locked balances
```

File: protos/sources/data-node/api/v2/trading_data.proto [hunk 4]
```diff
@@ -974,6 +981,18 @@ message PageInfo {
...
+message GetVestingBalancesSummaryRequest {
+  string party_id = 1 [(google.api.field_behavior) = REQUIRED];
+  optional string asset_id = 2;
+}
+
+message GetVestingBalancesSummaryResponse {
+  string party_id = 1;
+  optional uint64 epoch_seq = 2;
+  repeated vega.events.v1.PartyLockedBalance locked_balances = 3;
+  repeated vega.events.v1.PartyVestingBalance vesting_balances = 4;
```
Comment (candida-d) [inline protos/sources/data-node/api/v2/trading_data.proto:993]: docs please

File: protos/sources/vega/events/v1/events.proto [hunk 5]
```diff
@@ -13,6 +13,28 @@ import "vega/vega.proto";
...
+
+message PartyLockedBalance {
+  string asset = 1;
+  uint64 until_epoch = 2;
+  string balance = 3;
+}
+
+message PartyVestingBalance {
+  string asset = 1;
+  string balance = 2;
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:35]: docs here too please

File: datanode/api/trading_test.go [hunk 6]
```diff
@@ -172,6 +172,8 @@ func getTestGRPCServer(t *testing.T, ctx context.Context) (tidy func(), conn *gr
 	volumeDiscountStatsService := service.NewVolumeDiscountStats(sqlstore.NewVolumeDiscountStats(sqlConn))
 	volumeDiscountProgramService := service.NewVolumeDiscountPrograms(sqlstore.NewVolumeDiscountPrograms(sqlConn))
 	paidLiquidityFeesStatsService := service.NewPaidLiquidityFeesStats(sqlstore.NewPaidLiquidityFeesStats(sqlConn))
+	partyLockeBalances := service.NewPartyLockedBalances(sqlstore.NewPartyLockedBalances(sqlConn))
```
Comment (candida-d) [inline datanode/api/trading_test.go:175]: partyLockeBalances - locked?

File: datanode/api/trading_test.go [hunk 7]
```diff
@@ -226,6 +228,8 @@ func getTestGRPCServer(t *testing.T, ctx context.Context) (tidy func(), conn *gr
 		volumeDiscountStatsService,
 		volumeDiscountProgramService,
 		paidLiquidityFeesStatsService,
+		partyLockeBalances,
```
Comment (candida-d) [inline datanode/api/trading_test.go:231]: here too

File: protos/sources/data-node/api/v2/trading_data.proto [hunk 8]
```diff
@@ -974,6 +981,26 @@ message PageInfo {
   string end_cursor = 4;
 }
 
+// Request sent to list the vesting and locked balances for a party
+message GetVestingBalancesSummaryRequest {
+  // The party ID.
```
Comment (candida-d) [inline protos/sources/data-node/api/v2/trading_data.proto:986]:
```suggestion
  // Party ID to query vesting balances for.
```

File: protos/sources/data-node/api/v2/trading_data.proto [hunk 9]
```diff
@@ -974,6 +981,26 @@ message PageInfo {
...
+  string party_id = 1 [(google.api.field_behavior) = REQUIRED];
+  // An optional asset ID, all asset balances returned if not set.
```
Comment (candida-d) [inline protos/sources/data-node/api/v2/trading_data.proto:988]:
```suggestion
  // Optional asset ID, all asset balances returned if not set.
```

File: protos/sources/data-node/api/v2/trading_data.proto [hunk 10]
```diff
@@ -974,6 +981,26 @@ message PageInfo {
...
+  optional string asset_id = 2;
+}
+
+// The list of vesting and locked balances for a party
```
Comment (candida-d) [inline protos/sources/data-node/api/v2/trading_data.proto:992]:
```suggestion
// List of vesting and locked balances for a party.
```

File: protos/sources/data-node/api/v2/trading_data.proto [hunk 11]
```diff
@@ -974,6 +981,26 @@ message PageInfo {
...
+message GetVestingBalancesSummaryResponse {
+  // The party ID.
+  string party_id = 1;
+  // The epoch at which these balances are valid.
+  optional uint64 epoch_seq = 2;
+  // The list of locked balances for the party.
+  repeated vega.events.v1.PartyLockedBalance locked_balances = 3;
+  // The list of vesting balances for the party.
```
Comment (candida-d) [inline protos/sources/data-node/api/v2/trading_data.proto:1000]:
```suggestion
  // Party ID.
  string party_id = 1;
  // Epoch for which these balances are valid.
  optional uint64 epoch_seq = 2;
  // List of locked balances for the party.
  repeated vega.events.v1.PartyLockedBalance locked_balances = 3;
  // List of vesting balances for the party.
```

File: protos/sources/vega/events/v1/events.proto [hunk 12]
```diff
@@ -13,6 +13,42 @@ import "vega/vega.proto";
 
 option go_package = "code.vegaprotocol.io/vega/protos/vega/events/v1";
 
+// Summary of the vesting and locked balances for an epoch
+message VestingBalancesSummary {
+  // The epoch.
+  uint64 epoch_seq = 1;
+  // The parties summaries.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:20]:
```suggestion
  // Epoch for which these balances are valid.
  uint64 epoch_seq = 1;
  // Parties' summaries.
```

File: protos/sources/vega/events/v1/events.proto [hunk 13]
```diff
@@ -13,6 +13,42 @@ import "vega/vega.proto";
...
+}
+
+// A party vesting summary
+message PartyVestingSummary {
+  // The party ID.
+  string party = 1;
+  // A list of locked balances.
+  repeated PartyLockedBalance party_locked_balances = 2;
+  // A list of vesting balances.
+  repeated PartyVestingBalance party_vesting_balances = 3;
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:31]:
```suggestion
// Summary of a party's vesting balances
message PartyVestingSummary {
  // Party ID.
  string party = 1;
  // List of locked balances.
  repeated PartyLockedBalance party_locked_balances = 2;
  // List of vesting balances.
  repeated PartyVestingBalance party_vesting_balances = 3;
```

File: protos/sources/vega/events/v1/events.proto [hunk 14]
```diff
@@ -13,6 +13,42 @@ import "vega/vega.proto";
...
+}
+
+// A party locked balance for an asset
+message PartyLockedBalance {
+  // The locked asset.
+  string asset = 1;
+  // The epoch which will see the asset released.
+  uint64 until_epoch = 2;
+  // The locked balance.
+  string balance = 3;
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:41]:
```suggestion
// A party's locked balance for a given asset.
message PartyLockedBalance {
  // Asset ID.
  string asset = 1;
  // Epoch in which the balance will be released.
  uint64 until_epoch = 2;
  // Locked balance.
  string balance = 3;
```

File: protos/sources/vega/events/v1/events.proto [hunk 15]
```diff
@@ -13,6 +13,42 @@ import "vega/vega.proto";
...
+}
+
+// A balance being vested for a party
+message PartyVestingBalance {
+  // The asset ID.
+  string asset = 1;
+  // The balance being vested.
+  string balance = 2;
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:49]:
```suggestion
// Balance that is being vested for the party.
message PartyVestingBalance {
  // Asset ID.
  string asset = 1;
  // Balance that is vested.
  string balance = 2;
```

File: protos/sources/vega/events/v1/events.proto [hunk 16]
```diff
@@ -13,6 +13,42 @@ import "vega/vega.proto";
...
+
+// A party locked balance for an asset
+message PartyLockedBalance {
+  // The locked asset.
+  string asset = 1;
+  // The epoch which will see the asset released.
+  uint64 until_epoch = 2;
+  // The locked balance.
+  string balance = 3;
+}
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:42]: Is my change above correct - "epoch in which the **balance** will be released"?

**Review by candida-d — DISMISSED** [review]

**Review by EVODelavega — DISMISSED** [review]

**Review by EVODelavega — APPROVED** [review]

**Pull request merged**